File: frontend/src/features/owner/working-days.ts
// Weekday working pattern for daily-wage staff, mirroring the backend WorkingDays
// mask (Mon=bit0 .. Sun=bit6; 127 = every day of the week).

export const ALL_DAYS_MASK = 0b1111111; // 127

export const WEEKDAYS: { bit: number; label: string }[] = [
  { bit: 0, label: "Mon" },
  { bit: 1, label: "Tue" },
  { bit: 2, label: "Wed" },
  { bit: 3, label: "Thu" },
  { bit: 4, label: "Fri" },
  { bit: 5, label: "Sat" },
  { bit: 6, label: "Sun" },
];

export function normalizeMask(mask: number) {
  return mask <= 0 || mask > ALL_DAYS_MASK ? ALL_DAYS_MASK : mask;
}

export function hasDay(mask: number, bit: number) {
  return (normalizeMask(mask) & (1 << bit)) !== 0;
}

export function toggleDay(mask: number, bit: number) {
  return mask ^ (1 << bit);
}

export function weekdaysLabel(mask: number) {
  const effective = normalizeMask(mask);
  if (effective === ALL_DAYS_MASK) return "All week";
  const days = WEEKDAYS.filter((weekday) => hasDay(effective, weekday.bit)).map((weekday) => weekday.label);
  return days.length ? days.join(" ") : "No days";
}

// Number of working days that fall in the current calendar month.
export function workingDaysInCurrentMonth(mask: number) {
  const effective = normalizeMask(mask);
  const now = new Date();
  const daysInMonth = new Date(now.getFullYear(), now.getMonth() + 1, 0).getDate();
  let count = 0;
  for (let day = 1; day <= daysInMonth; day++) {
    const jsDay = new Date(now.getFullYear(), now.getMonth(), day).getDay(); // 0=Sun..6=Sat
    const bit = jsDay === 0 ? 6 : jsDay - 1; // map to Mon=0..Sun=6
    if (hasDay(effective, bit)) count++;
  }
  return count;
}

File: src/main/java/com/khatiyan/d_modules/staff/model/WorkingDays.java
package com.khatiyan.d_modules.staff.model;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.YearMonth;

/**
 * Weekday working pattern for daily-wage staff, encoded as a 7-bit mask where
 * bit {@code (DayOfWeek.getValue() - 1)} is set when the worker is on duty that
 * day — Monday = bit 0 ... Sunday = bit 6. {@link #ALL_DAYS} (127) means every
 * day of the week. Only meaningful for DAILY staff; monthly staff keep it at
 * {@code ALL_DAYS}.
 */
public final class WorkingDays {

    public static final int ALL_DAYS = 0b1111111; // 127

    private WorkingDays() {
    }

    /** Clamps an incoming mask to a valid 1..127 range, defaulting to all days. */
    public static int normalize(int mask) {
        return (mask <= 0 || mask > ALL_DAYS) ? ALL_DAYS : mask;
    }

    public static boolean worksOn(int mask, DayOfWeek day) {
        return (normalize(mask) & (1 << (day.getValue() - 1))) != 0;
    }

    /** Number of working days that fall in the given calendar month. */
    public static int countInMonth(int mask, YearMonth month) {
        int effective = normalize(mask);
        int count = 0;
        LocalDate day = month.atDay(1);
        LocalDate end = month.atEndOfMonth();
        while (!day.isAfter(end)) {
            if ((effective & (1 << (day.getDayOfWeek().getValue() - 1))) != 0) {
                count++;
            }
            day = day.plusDays(1);
        }
        return count;
    }
}
